### PR TITLE
storage: add merge queue

### PR DESCRIPTION
**Pull request by benesch**
Add a merge queue, analogous to the split queue, that continually looks
for ranges that are below the minimum size threshold and merges them
with their right-hand neighbor.

---

I want to try this on at least the clear range roachtest before merging, but figured I'd get it out for early review.

### CODE REVIEW

**Review by bdarnell — APPROVED** [review]
> Reviewed 8 of 8 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27733)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_merge_test.go, line 1135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHnx12i5iAk9rQCAR1A:-LHnx12i5iAk9rQCAR1B:b-o15tr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e60ba10a4841e4d667c2d4b90456290934894ac1/pkg/storage/client_merge_test.go#L1135)):*
> ```Go
> }
> 
> func TestRangeStatsRequest(t *testing.T) {
> ```

Put this in `replica_test.go` instead of client_merge_test.go (unless it can't go into package storage for some reason that I'm not seeing). 
___
*[pkg/storage/merge_queue.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHnzETvElpAmSBWZp7D:-LHnzETw0nZmuN1dnZwG:b-7iuys) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L35)):*
> ```Go
> const (
> 	// mergeQueueTimerDuration is the duration between merges of queued ranges.
> 	mergeQueueTimerDuration = 0 // zero duration to process merges greedily.
> ```

I think we want a non-zero duration here (we use zero durations for queue like replicate and split that can cause availability problems if they can't keep up). 
___
*[pkg/storage/merge_queue.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo-02ZAHlvolcfSrb6:-LHo-02ZAHlvolcfSrb7:biczm2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L55)):*
> ```Go
> 		"kv.range_merge.max_rhs_size",
> 		"maximum size of the right-hand range in a merge",
> 		0,
> ```

Even "empty" ranges contain a range descriptor, stats records, etc. Are all of those tracked separately so ranges will reach zero bytes by the relevant metric?
___
*[pkg/storage/merge_queue.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo-Ukw4uidiw1AUURt:-LHo-Ukw4uidiw1AUURu:b-50s9gq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L80)):*
> ```Go
> // merges can only proceed if a) the right-hand neighbor is smaller than
> // MergeMaxRHSSize, and b) the merged range would not need to be immediately
> // split, e.g. because the new range would exceed the maximum size threshold.
> ```

Another reason why a merged range might immediately need to split is if it spans two table IDs. We should make sure that this is tested and not just the size threshold case. 
___
*[pkg/storage/merge_queue.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo06EoEqJQsXOez50z:-LHo06EoEqJQsXOez51-:bt0nx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L152)):*
> ```Go
> 
> 	if desc.EndKey.Equal(roachpb.RKeyMax) {
> 		// The last range cannot be merged away.
> ```

s/cannot be merged away/has no right-hand neighbor to merge with/
___
*[pkg/storage/merge_queue.go, line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo7IJ0BHNBw0yZsm19:-LHo7IJ0BHNBw0yZsm1A:b-awn525) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L224)):*
> ```Go
> 		return err
> 	}
> 	if maxBytes := MergeMaxRHSSize.Get(&mq.store.ClusterSettings().SV); rhsStats.Total() > maxBytes {
> ```

Maybe this should use the smaller of the cluster setting and the zone config minimum. Once we're no longer limited to merging small ranges, the rule will be that both sides need to be below the zone config minimum (and the RHS setting will be phased out), right?
___
*[pkg/storage/merge_queue.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo8zQwB3t95TRTEDG2:-LHo8zQwB3t95TRTEDG3:bmuzov4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L225)):*
> ```Go
> 	}
> 	if maxBytes := MergeMaxRHSSize.Get(&mq.store.ClusterSettings().SV); rhsStats.Total() > maxBytes {
> 		log.VEventf(ctx, 2, "skipping merge: right-hand neighbor exceeds size threshold %d with %d bytes",
> ```

If we reach this point, we have an LHS under the zone config minimum size, and a non-empty RHS. These conditions will probably still hold the next time this range is queued, so we'll keep doing the RangeStats call every time the range is queued. I think it may be better to treat this like the replica GC queue (just run it with a long cycle time) instead of trying to use the size of the LHS to limit how many RangeStats calls we make.

**Review by benesch — COMMENTED** [review]
Ok, I addressed most of your feedback, @bdarnell. I think your proposal to always send RangeStats requests, even if the LHS is oversize, makes sense, but I'd like to leave it to another PR if that's alright with you. I'd love to get this merged so we can start running roachtests with the merge queue enabled.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-:-LHobR4c184O8l6--l7R:b-3ymyeu)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_merge_test.go, line 1135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHnx12i5iAk9rQCAR1A:-LHoZXmi91Dw8MsVmZmt:b-mcgekv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e60ba10a4841e4d667c2d4b90456290934894ac1/pkg/storage/client_merge_test.go#L1135)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Put this in `replica_test.go` instead of client_merge_test.go (unless it can't go into package storage for some reason that I'm not seeing). 
</blockquote></details>

Done. I just needed to move `writeRandomDataToRange` to a file shared between the test and xtest packages.
___
*[pkg/storage/merge_queue.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHnzETvElpAmSBWZp7D:-LHoV4LwAGBiij3oEwFY:b1k7qdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L35)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we want a non-zero duration here (we use zero durations for queue like replicate and split that can cause availability problems if they can't keep up). 
</blockquote></details>

I agree, but I also think it's useful to see just how disruptive greedy merges are during the next few weeks. I've filed a TODO and an issue (#27769) to set it higher before release.
___
*[pkg/storage/merge_queue.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo-02ZAHlvolcfSrb6:-LHoPriH7kEegLNFH_A0:b9jsvkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L55)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Even "empty" ranges contain a range descriptor, stats records, etc. Are all of those tracked separately so ranges will reach zero bytes by the relevant metric?
</blockquote></details>

I believe so. `mvccStats.Total()` is `KeyBytes + ValBytes`, which is only user data, but all versions of user data. So a range will become empty once all user data is deleted and the data has been GC'd. That seems right.
___
*[pkg/storage/merge_queue.go, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo-Ukw4uidiw1AUURt:-LHoVg55C4sW2_uZjTRP:b-ja9j70) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L80)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Another reason why a merged range might immediately need to split is if it spans two table IDs. We should make sure that this is tested and not just the size threshold case. 
</blockquote></details>

That case is covered in the ShouldQueue tests. Do you think it needs to be more tested than that?
___
*[pkg/storage/merge_queue.go, line 152 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo06EoEqJQsXOez50z:-LHoWHQ_3ZP7x48XFYHJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L152)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/cannot be merged away/has no right-hand neighbor to merge with/
</blockquote></details>

Done.
___
*[pkg/storage/merge_queue.go, line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo7IJ0BHNBw0yZsm19:-LHob2Ru-Y1jUv1qnNyL:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L224)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe this should use the smaller of the cluster setting and the zone config minimum. Once we're no longer limited to merging small ranges, the rule will be that both sides need to be below the zone config minimum (and the RHS setting will be phased out), right?
</blockquote></details>

Good point. Done.
___
*[pkg/storage/merge_queue.go, line 225 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27733#-LHo8zQwB3t95TRTEDG2:-LHoZhxR9DGtgTMdk1dE:b-84ojuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d78a73418d12baf9a1ec4272c33128b092a865d/pkg/storage/merge_queue.go#L225)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If we reach this point, we have an LHS under the zone config minimum size, and a non-empty RHS. These conditions will probably still hold the next time this range is queued, so we'll keep doing the RangeStats call every time the range is queued. I think it may be better to treat this like the replica GC queue (just run it with a long cycle time) instead of trying to use the size of the LHS to limit how many RangeStats calls we make. 
</blockquote></details>

Yeah, that sounds reasonable. We could also move to an entirely event driven model where ranges are only queued when their size drops beneath the threshold.

I'm going to leave as-is for now in the interest of getting something in that we can test nightly.

**Review by bdarnell — APPROVED** [review]
> Reviewed 2 of 11 files at r3, 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27733)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by benesch** [comment]
`TestMergeQueue` was failing due to a merge deadlock that's tracked #27820. Marking it as skipped for now so I can land this PR. I'll unskip it when I fix #27820.

**Comment by benesch** [comment]
bors r=bdarnell

**Comment by benesch** [comment]
bors r=bdarnell

**Pull request merged**
